### PR TITLE
net: ipv6: Check the size of the ICMPv6 packet

### LINKED ISSUE #10970: net: ipv6: Zephyr replies to malformed ICMPv6 message
Malformed messages should be dropped instead.

[icmp-trunc.pcap.gz](https://github.com/zephyrproject-rtos/zephyr/files/2534738/icmp-trunc.pcap.gz)

### PR DESCRIPTION
**Pull request by jukkar**
The minimum size is 8 bytes, drop packet if shorter.

Fixes #10970

Signed-off-by: Jukka Rissanen <jukka.rissanen@linux.intel.com>

### CODE REVIEW

File: subsys/net/ip/ipv6.c [hunk 1]
```diff
@@ -159,6 +159,15 @@ static inline enum net_verdict process_icmpv6_pkt(struct net_pkt *pkt,
 		return NET_DROP;
 	}
 
+#define MIN_ICMPV6_LEN (1 + 1 + 2 + 2 + 2) /* type, code, checksum,
```
Comment (tbursztyka) [inline subsys/net/ip/ipv6.c]: Having such macro here is not nice, could you put it in icmpv6.h, have a name prefixed NET_ICMPV6_ and reuse also NET_ICMPV6H_LEN for the hdr part (type, code, chksum).

**Review by tbursztyka — CHANGES_REQUESTED** [review]

**Comment by jukkar** [comment]
Fixing the check, the proposed one was not doing things properly.

**Comment by jukkar** [comment]
Fixed the payload_len check once more.

**Review by rmstoi — APPROVED** [review]

File: subsys/net/ip/icmpv6.c [hunk 2]
```diff
@@ -304,13 +304,18 @@ static enum net_verdict handle_echo_request(struct net_pkt *orig)
 
 	iface = net_pkt_iface(orig);
 
+	payload_len = ntohs(NET_IPV6_HDR(orig)->len) -
+		net_pkt_ipv6_ext_len(orig) - NET_ICMPH_LEN;
+	if (payload_len < NET_ICMPV6_UNUSED_LEN) {
```
Comment (pfalcon) [inline subsys/net/ip/icmpv6.c]: I'd say this whole chunk of code is not the clearest one. `payload_len` doesn't seem to be used elsewhere. And then, I'd say we should calculate `icmp_pkt_len`, and check that it's no smaller than `NET_ICMPH_LEN + NET_ICMPV6_UNUSED_LEN`. Not insisting though, it's just going thru those steps is what took me to verify that this patch looks correct.

Comment (jukkar) [reply subsys/net/ip/icmpv6.c]: The payload_len was not used earlier for some reason. Now it holds the length of the ICMP messge, and if it is too small, then we can safely drop it. In ICMP protocol, the first 4 bytes are mandatory, the latter 4 (the NET_ICMPV6_UNUSED_LEN define) depends on what packet type is being sent. For echo-req, the latter four bytes contain the sequence and identifier numbers.

Comment (jukkar) [reply subsys/net/ip/icmpv6.c]: As there can be user payload data in icmp message, the payload_len is kind of descriptive in this case.

**Review by tbursztyka — APPROVED** [review]

**Pull request merged**